File: df_imoveis.py
import time
import json
import csv
import re
from urllib.parse import urljoin, urlparse
import requests
from bs4 import BeautifulSoup

BASE = "https://www.dfimoveis.com.br"
LIST_URL = ("https://www.dfimoveis.com.br/venda/df/brasilia/asa-norte,asa-sul/imoveis/"
            "3,4-quartos?suites=1&vagasdegaragem=1&valorfinal=1200000&areainicial=95")

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7",
}

PRICE_RX = re.compile(r"R\$\s*[\d\.]{1,3}(?:\.\d{3})*(?:,\d{2})?", re.I)

def get_soup(url: str) -> BeautifulSoup:
    r = requests.get(url, headers=HEADERS, timeout=30)
    r.raise_for_status()
    return BeautifulSoup(r.text, "html.parser")

def collect_listing_links(list_url: str) -> list[str]:
    """
    Pega todos os links que parecem ser de imóveis (rota /imovel/).
    Evita duplicados e links externos.
    """
    soup = get_soup(list_url)
    links = set()
    for a in soup.select("a[href]"):
        href = a.get("href", "")
        if "/imovel/" in href:
            full = urljoin(BASE, href)
            # garante domínio correto
            if urlparse(full).netloc.endswith("dfimoveis.com.br"):
                links.add(full)
    return sorted(links)

def extract_text_block(soup: BeautifulSoup) -> str:
    """
    Extrai a descrição do imóvel procurando por uma div com classe 'imovel-title'
    e dentro dela um h2 que contém a descrição.
    """
    # Procura a div com a classe imovel-title
    imovel_title_div = soup.find("div", class_="imovel-title")
    if imovel_title_div:
        # Dentro dessa div, procura por um h2
        h2_desc = imovel_title_div.find("h2")
        if h2_desc:
            desc_text = h2_desc.get_text(" ", strip=True)
            if desc_text:
                return desc_text
    
    # Fallback: tenta a seção 'Descrição'
    desc_header = soup.find(lambda t: t.name in ("h2", "h3", "h4") and "descri" in t.get_text(strip=True).lower())
    if desc_header:
        # texto do(s) próximo(s) irmãos até um limite
        texts = []
        for sib in desc_header.next_siblings:
            if getattr(sib, "name", None) in ("h2", "h3", "h4"):
                break
            txt = getattr(sib, "get_text", lambda **k: "")(strip=True)
            if txt:
                texts.append(txt)
            if len(" ".join(texts)) > 1200:
                break
        if texts:
            return " ".join(texts)

    # Fallback 2: tenta metadados/resumo no topo (ex.: título + subtítulo)
    pieces = []
    title = soup.find(["h1", "h2"])
    if title:
        pieces.append(title.get_text(" ", strip=True))
    sub = soup.find(["h3", "p"])
    if sub:
        pieces.append(sub.get_text(" ", strip=True))
    body = soup.find("main") or soup
    if body:
        chunk = body.get_text(" ", strip=True)
        if chunk:
            pieces.append(chunk)
    # limpa e reduz
    flat = " ".join(pieces)
    flat = re.sub(r"\s+", " ", flat)
    return flat[:2000]

def extract_price(soup: BeautifulSoup) -> str | None:
    """
    Extrai o preço do imóvel procurando por uma div com classe 'imovel-price'
    e dentro dela elementos h4 que contenham valores parseáveis como números.
    """
    # Procura a div com a classe imovel-price
    imovel_price_element = soup.find("h4", class_="precoAntigoSalao")
    if imovel_price_element:
        return imovel_price_element.get_text(" ", strip=True)
            
          

def parse_property(url: str) -> dict:
    soup = get_soup(url)
    price = extract_price(soup)
    desc = extract_text_block(soup)

    return {
        "titulo": desc,
        "valor": price or "",
        "link": url,
    }

def collect_df_imoveis_properties():
    links = collect_listing_links(LIST_URL)
    if not links:
        print("Nenhum link de imóvel encontrado na listagem.")
        return

    resultados = []
    for i, link in enumerate(links, 1):
        try:
            item = parse_property(link)
            resultados.append(item)
            print(f"[{i:02d}] {item['titulo']}\n    {item['valor']} | {item['link']}\n")
        except Exception as e:
            print(f"Erro ao processar {link}: {e}")
        time.sleep(1.2)  # educação com o site
    return resultados

File: src/df_imoveis.ts
import { delay } from "@std/async/delay";
import type { CheerioAPI } from "cheerio";

import { Property } from "./types.ts";
import {
  buildAbsoluteUrl,
  fetchDocument,
  isSameDomain,
  normalizeWhitespace,
  printProperty,
} from "./utils.ts";

const BASE = "https://www.dfimoveis.com.br";
const LIST_URL =
  "https://www.dfimoveis.com.br/venda/df/brasilia/asa-norte,asa-sul/imoveis/3,4-quartos?suites=1&vagasdegaragem=1&valorfinal=1200000&areainicial=90";

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
  "Accept":
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
  "Accept-Language": "pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7",
  "Accept-Encoding": "gzip, deflate, br",
  "Referer": "https://www.dfimoveis.com.br/",
  "Connection": "keep-alive",
  "Sec-Fetch-Dest": "document",
  "Sec-Fetch-Mode": "navigate",
  "Sec-Fetch-Site": "same-origin",
  "Sec-Fetch-User": "?1",
  "Upgrade-Insecure-Requests": "1",
  "Cache-Control": "max-age=0",
};

async function getDocument(url: string): Promise<CheerioAPI> {
  return await fetchDocument(url, HEADERS);
}

async function collectListingLinks(listUrl: string): Promise<string[]> {
  const $ = await getDocument(listUrl);
  const links = new Set<string>();

  $("a[href]").each((_, element) => {
    const href = $(element).attr("href") ?? "";
    if (!href.includes("/imovel/")) {
      return;
    }

    const absolute = buildAbsoluteUrl(BASE, href);
    if (isSameDomain(absolute, "dfimoveis.com.br")) {
      links.add(absolute);
    }
  });

  return Array.from(links).sort();
}

function extractTextBlock($: CheerioAPI): string {
  const headline = normalizeWhitespace($("div.imovel-title h2").first().text());
  if (headline) {
    return headline;
  }

  const descHeader = $("h2, h3, h4")
    .filter((_, element) =>
      normalizeWhitespace($(element).text()).toLowerCase().includes("descri")
    )
    .first();

  if (descHeader.length) {
    const texts: string[] = [];
    $(descHeader)
      .nextAll()
      .each((_, element) => {
        const tag = element.tagName?.toLowerCase();
        if (tag && ["h2", "h3", "h4"].includes(tag)) {
          return false;
        }

        const text = normalizeWhitespace($(element).text());
        if (text) {
          texts.push(text);
        }

        if (texts.join(" ").length > 1200) {
          return false;
        }

        return undefined;
      });

    if (texts.length > 0) {
      return normalizeWhitespace(texts.join(" "));
    }
  }

  const pieces: string[] = [];
  const title = normalizeWhitespace($("h1, h2").first().text());
  if (title) {
    pieces.push(title);
  }

  const subtitle = normalizeWhitespace($("h3, p").first().text());
  if (subtitle) {
    pieces.push(subtitle);
  }

  const body = normalizeWhitespace($("main").text() || $.root().text());
  if (body) {
    pieces.push(body);
  }

  return normalizeWhitespace(pieces.join(" ")).slice(0, 2000);
}

function extractPrice($: CheerioAPI): string {
  const price = normalizeWhitespace($("h4.precoAntigoSalao").first().text());
  return price;
}

async function parseProperty(url: string): Promise<Property> {
  const $ = await getDocument(url);
  const price = extractPrice($);
  const description = extractTextBlock($);

  return {
    titulo: description,
    valor: price,
    link: url,
  };
}

export async function collectDfImoveisProperties(): Promise<Property[]> {
  try {
    const links = await collectListingLinks(LIST_URL);
    if (links.length === 0) {
      console.log("Nenhum link de imóvel encontrado na listagem.");
      return [];
    }

    const results: Property[] = [];
    for (const link of links) {
      try {
        const property = await parseProperty(link);
        results.push(property);
        printProperty(property);
      } catch (error) {
        console.error(`Erro ao processar ${link}:`, error);
      }

      await delay(1200);
    }

    return results;
  } catch (error) {
    if (error instanceof Error && error.message.includes("403")) {
      console.error(
        "⚠️  AVISO: DF Imóveis bloqueou o acesso (403 Forbidden).",
      );
      console.error(
        "   Isso é comum em ambientes de CI/CD como GitHub Actions.",
      );
      console.error("   O crawler continuará apenas com os outros sites.\n");
    } else {
      console.error("❌ Erro ao coletar imóveis do DF Imóveis:", error);
    }
    return [];
  }
}
